Add support for Nanoleaf Essentials / Replace aionanoleaf through aionanoleaf2

loebi-ch opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The current Nanoleaf integration depends on the abandoned and not maintained PyPi package `aionanoleaf` that doesn't work for the newer **Nanoleaf Essentials** devices (that changed their API slightly) and is also missing the emersion modes for the Nanoleaf **4D Screen Mirroring** kit. All pull requests and contact trials failed. Therefore I created the new PyPi package `aionanoleaf2` which fixes the above two problems/features while keeping the backwards compatibility.

This pull request replaces the dependency from `aionanoleaf` with `aionanoleaf2` and allows to select the emersion modes (1D, 2D, 3D and 4D) from the effects list for Nanoleaf Screen Mirroring devices.

A diff between `aionanoleaf` and `aionanoleaf2` can be found here: https://github.com/milanmeu/aionanoleaf/pull/18/files

I have changed the codeowners to @loebi-ch (me), @JaspervRijbroek (contributor of Essentials fix) and @jonathanrobichaud4 (contributor of Screen Mirroring feature) because the other two guys haven't responded to pull requests for years.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issues: fixes https://github.com/home-assistant/core/issues/132084, https://github.com/home-assistant/core/issues/152849, https://github.com/home-assistant/core/issues/152713
- This PR is related to issue: https://github.com/home-assistant/core/issues/142960, https://github.com/home-assistant/core/issues/141964 (This two issues will be probably solved by the pull request and the update of the device firmware)
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

loebi-ch committed:
Add support for Nanoleaf Essentials and Screen Mirroring

loebi-ch committed:
Merge branch 'dev' into dev

joostlek commented:
Did I hit a year of codeownership yet? Idk. Let me check with Milan

loebi-ch commented:
> Did I hit a year of codeownership yet? Idk. Let me check with Milan

Yes please check with Milan. Lots of Nanoleaf issues are just getting stuck because of `aionanoleaf`. I'm not keen on taking the lead, but someone has to.

loebi-ch commented:
Hi @sirmeili, @milanmeu and @joostlek

I have created a custom component that can be used with HACS as a custom repository.
You can find it here: https://github.com/loebi-ch/nanoleaf

### The story behind HACS and the name conflict `nanoleaf` vs `nanoleaf2`

I tried to put that component into HACS default repository, but unfortunately that's not possible if it's a successor of a core component. The only way to use it with HACS is if you include it in HACS as described in the [README](https://github.com/loebi-ch/nanoleaf/blob/main/README.md) as a custom repository.

First I created a repository called `nanoleaf`, but struggled with the branding that is necessary to make a PR for HACS, so I created another repository called `nanoleaf2`. With this repository I achieved that the branding was accepted, but the PR to HACS default has been rejected due to [policies](https://hacs.xyz/docs/publish/include/):

> Custom integrations used for alpha or beta testing core integrations, or those that override core integrations, are not accepted as defaults. You can still use them as a [custom repository](https://hacs.xyz/docs/faq/custom_repositories/).

Furthermore, the name of the custom component has to be the same, so that the core component is replaced correctly.

So in the end I deleted both my repositories `nanoleaf` and `nanoleaf2`, and then created the new one:


### The final custom component

https://github.com/loebi-ch/nanoleaf

So everybody can go with this repository and install it with little effort to make Nanoleaf Essentials devices work.

Of course, it would be nice if this repository, or at least my changes, made it into the official Home Assistant release, but with this custom component, you now have an easy way to get your devices working within Home Assistant. Light up your homes!


I wish you a happy Advent season.
Andy

joostlek commented on homeassistant/components/nanoleaf/manifest.json:
```diff
@@ -1,15 +1,17 @@
   "domain": "nanoleaf",
   "name": "Nanoleaf",
-  "codeowners": ["@milanmeu", "@joostlek"],
+  "version": "2.0.0",
```
Please remove the version

joostlek reviewed (changes requested):
Okay so we gave enough time, maybe I will see Milan next weekend. In any case, let's start this migration. Ideally the first PR is just the change of the library and changes needed to support the new library. New features should go in a later PR.

Also, it would be cool if all 3 new codeowners send me a message on Discord so we can add you to the right channels

loebi-ch committed:
Modify Nanoleaf manifest.json

Keep previous codeowners.
Version removed.
Exact requirement of `aionanoleaf2` in version 1.0.2.

loebi-ch committed:
Update CODEOWNERS for nanoleaf component owners

Keep previous codeowners.

loebi-ch committed:
Update aionanoleaf2 version to 1.0.2

loebi-ch committed:
Remove emersion models and modes from const.py

Removed emersion models and modes from constants.
We insert these in a separate PR.

loebi-ch committed:
Remove 4D Screen Mirroring feature (emersion models)

joostlek committed:
Merge branch 'dev' into loebi-ch/dev

# Conflicts:
#	CODEOWNERS
#	homeassistant/components/nanoleaf/manifest.json

joostlek reviewed (approved):
If @JaspervRijbroek and @jonathanrobichaud4 are also able to send me a message on Discord, that'd be great!

joostlek committed:
Merge branch 'dev' into dev

joostlek committed:
Fix

joostlek committed:
Merge branch 'dev' of https://github.com/loebi-ch/home-assistant-nanoleaf-upgrade into loebi-ch/dev
